Ports soft lights featured in other servers

Datraen opened the pull request:
Ports PolarisSS13/Polaris/pull/3286 with changes appropriate for the current codebase.

5/8/17
![](http://i.imgur.com/e2ryo21.png)

Older:
http://i.imgur.com/0iPaUuz.png
http://i.imgur.com/MP28ydZ.png
http://i.imgur.com/TFwL2Vq.png
http://i.imgur.com/XpEHQxq.png
http://i.imgur.com/YZ1K51f.png
http://i.imgur.com/dmhOoZh.png

Leshana committed:
Ports Soft Lighting implementation from Polaris.

/vg/station, /tg/station, goon and Paradise code with edits to ensure compatability with current Bay code.

New() returning
utilizes handle_opacity_change for entering/exiting turfs.
Removes the override for set_opacity, as it contained a lot of redundant code. Non-redundant code moved to the other proc.
Fixes light bleeding
Corrects cherry-picking not having set_opacity() for smoke, it did in cherry-picked PR.

[7b8036e] Corrects cherry picking errors and updates maps for soft lighting.

[bb31c5a] Allow set_opacity() to work for turfs too.

Fixes lighting opacity not updating on mined rock walls and false doors. (Not relevant for Baystation)

[9060542] Restores "soft" darksight vision.

The previous client-side optimization made lighting overlays go fully
opaque when it was dark, making it so species darksight didn't let you see
anything.  This fixes that by making a dark_soft icon state that has
opacity matching the LIGHTING_SOFT_THRESHOLD.

[9256758] Optimizes lighting overlays for slow clients.

Ports https://github.com/tgstation/tgstation/pull/25778 and https://github.com/tgstation/tgstation/pull/25861
* Instead of using color matrix for all overlays, fully dark and fully light use static black and transparant icon states without color matrix.
* Also implements LIGHTING_ROUND_VALUE so that lighting values actually will reach 0 or 1.

[3f3d0ab] Replace direct access of lighting overlays with T.get_lumcount()

[3eb81b2] Port tg/paradise/good soft-edge lighting

Ports https://github.com/ParadiseSS13/Paradise/pull/6161
Its merged in nearly verbatim as far as the lighting module goes.
Changes outside the lighting module are merged based on our codebase.

[3c8bf01] Replace `opacity = X` with `set_opacity(X)` so that the lighting system can update properly.

Chiquiii commented:
You are doing gods work. The only worry is this is gonna make the server shit itself.

AtomicWorm commented:
Bless you man. Bless you.

mustafakalash commented:
This is pretty ugly. It doesn't match the hard-edged and pixelated style of space mans.

Chiquiii commented:
And light squares aren't ugly?

PJB3005 commented:
👌

redstryker commented:
Soft light is *much* better than hard light.

chaoko99 commented:
**I** **T** **'** **S** **L** **I** **T** 

Someone smack me.

Minijar commented:
Perhaps it's just me, but it looks like there's no coloured ambient lighting from objects in your screenshots? Ie, green light from fire alarms or APCs.

chaoko99 commented:
There's just one, right in the corner of one of the example images, the last one.

Datraen commented:
@Minijar 
![](http://i.imgur.com/dmhOoZh.png)
Added to OP as well.

ghost commented:
Can you now please also remove ugly alert area overlay and make nearby lighting fixtures coloring to red?

ThatOneGuy42 commented:
Yes daddy please

On Apr 27, 2017 5:12 PM, "Quardbreak" <notifications@github.com> wrote:

> Can you now please also remove ugly alert area overlay and make nearby
> lighting fixtures coloring to red?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Baystation12/Baystation12/pull/17044#issuecomment-297853678>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYOUaeqK0DrXrh_8kyPD7KcAgEbyWh8ks5r0RLPgaJpZM4NKmrg>
> .
>

Datraen commented:
What do you mean by "ugly alert area overlay" ?

Chiquiii commented:
he means the red flashing lights over hallways when evacuating or something. Maybe with the fire alarms as well, im not sure.

Nerev4r commented:
Note that this PR also makes alien darksight actually work.

Zero-Bits commented:
Yes please. This would be so sweet.

mattmatterson111 commented:
It's glorious.

Datraen commented:
I would prefer to do that with changing to a light configuration, similar to low power lights. It's an overlay put on by the area, rather than a light event.

Worth noting that there is some tweaking to be done on the map and lights in the code to make the lights nice, and rooms go dark a bit better when someone turns off the lights. I wanted to make this before making large map changes that would conflict.

chaoko99 commented:
Can you please post a screen from the airlock into maint by the aux. head and the saferoom on deck 1? I always thought that was a good spot of lighting.

Datraen commented:
http://i.imgur.com/ucG3BIH.png
Posting more embedded pictures will make this page larger than it has to be.

BlueNexus commented:
Oh hell yes!

Minijar commented:
What's the impact client side if any?

PsiOmegaDelta commented:
So many loops being called over and over again...

Datraen commented:
I'm not sure what the impact client-side is, as I was hosting it on my desktop and the server usually runs slowly. The only other computer I have access to and complete control over is running an athlon 2 x4 cpu, which would have the same issue of server side lag.

Chiquiii commented:
We could run play tests?

ghost commented:
Can you check starlight option?

NullSnapshot commented:
👍 on starlight support check. I haven't seen any goon-like lighting using starlight.
Also if @Baystation12/byond-game-devs think it's warranted we can flag this for playtesting for performance evaluation. Considering it's a refactor of the lighting system I'm tempted to say yes for a playtest.

PJB3005 commented:
@NullSnapshot TG and Citadel have starlight plus /vg/lights.

It works and it looks beautiful too when set to a really dim amount of light.

PsiOmegaDelta commented on code/modules/lighting/lighting_area.dm:
```diff
@@ -0,0 +1,9 @@
+	var/dynamic_lighting = TRUE
+
+/area/New()
+	. = ..()
```
According to the word of God (Lummox) New() should never return.

Datraen commented:
If we're working on playtesting it means that I have to work on getting map updates and tweaks on the lights ready, so there aren't rooms that are fullbright and lights leaking through from one room almost illuminating half of the other.

I'll get it done when I'm able, it's something that I wanted to do either a) with a maplock or b) after this was merged so I don't have to do ~fun~ things to avoid merge conflicts.

Datraen commented:
@NullSnapshot Starlight is a config option, at starlight 3 we get http://i.imgur.com/uju1xwp.png

Anticept commented:
Keep in mind goonlighting (the original creators of this soft lighthing are goon), are an ENORMOUS strain on resources. On paradise, it eclipses even a 10x sped up linda.

This has been reduced a reasonable amount though with byond 511, which resolved a problem with the color matrix cache which caused a memleak and CPU rise.

PJB3005 commented:
@Anticept which is why you give it a low subsystem priority so it doesn't hog CPU. Issue is bay doesn't have StonedMC yet so it's not doable.

Also, on Paradise lighting processes as fast as it can anyways.

> (the original creators of this soft lighthing are goon)

Yes but no code has been taken, also goon's lighting (robustlight in the code) has some stuff like oversaturation, thus to avoid confusion I call it /vg/lights.

PJB3005 commented:
Also this is my sweet spot for starlight strength: https://cloud.githubusercontent.com/assets/10257081/20033355/ee9e8f6a-a374-11e6-9c65-5965fa843aa3.png

BlueNexus commented:
I suggest waiting until we're finished porting StonedMC before testing/merging this.

PurpleMartinJCK commented:
In https://github.com/Datraen/Baystation12/blob/a8e1041fbfa2cf82d47f76ee07723c16fe169434/code/modules/mob/living/carbon/human/life.dm

With the addition of the piece to line 598 the lines 593-597 aren't needed anymore. 

Do you know if T.get_lumcount() will give comparable values as the previous method in similar light value amounts?

PJB3005 commented:
@FTangSteve yeah should be good.

Datraen commented:
@FTangSteve I thought I cut that out, I'll revise that momentarily.

I have a feeling that this PR will get a "requires squash" by the time I'm done.

PurpleMartinJCK commented:
Yeah no problem! I might grab a copy and do some balancing with the light and nutrition before releasing if I have the time

StarSmasher44 commented:
Would probably work better in combination with the subsystem processor PR that is still being worked on

PsiOmegaDelta commented on code/modules/lighting/lighting_turf.dm:
```diff
@@ -1,37 +1,97 @@
+	if(Obj && Obj.opacity)
+		opaque_counter++
+		if(!old_counter && opaque_counter)
+			reconsider_lights()
```
You also want to register to the opacity event of any objects that enter a turf (whether they're originally solid or not) to correctly update the opacity counter, as well as un-register the event when they exit or are destroyed.

Anticept commented:
There is one other thing as well.

If someone has out of date dx9 libs, it will cause a rainbow tile effect. It should be noted that anyone who gets this problem or huge lag that didn't exist before needs to run the dx install from https://www.microsoft.com/en-us/Download/confirmation.aspx?id=35

chaoko99 commented:
Very dumb request: Irradiate a promethean, I wonder how it looks.

Also, are we able to adjust light falloff?


On Tue, May 2, 2017 at 12:04 AM, CoreyLee Hassell <notifications@github.com>
wrote:

> There is one other thing as well.
>
> If someone has out of date dx9 libs, it will cause a rainbow tile effect.
> It should be noted that anyone who gets this problem needs to run the dx
> install fromhttps://www.microsoft.com/en-us/Download/confirmation.
> aspx?id=35
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Baystation12/Baystation12/pull/17044#issuecomment-298515172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfPBEGjSXvH_reu4x2ve6hsoKl_I3pzks5r1tVygaJpZM4NKmrg>
> .
>

Datraen commented:
Light falloff is the range of the light. So yes?

Datraen commented:
Light bleeding bug was fixed, DNM remains only for tweaking.

chaoko99 commented:
Rather, its falloff curve. I dunno how to describe it past graphing a 2^2
vs 2^3

In that I mean "Incredibly bright on the same turf, but sharply darker as it goes, or perhaps the inverse.
On Wed, May 3, 2017 at 6:57 PM, Datraen <notifications@github.com> wrote:

> Light bleeding bug was fixed, DNM remains only for tweaking.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Baystation12/Baystation12/pull/17044#issuecomment-299081305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEfPBIzITH7dkqOwV8OrA2mxmfHiBsbPks5r2TBxgaJpZM4NKmrg>
> .
>

PsiOmegaDelta commented on code/modules/lighting/lighting_atom.dm:
```diff
@@ -41,36 +44,27 @@
 
+	if(opacity && isturf(loc))
+		var/turf/T = loc
+		T.Entered(src)
```
Probably want to execute ```Entered()``` whether or not opacity is set, to register to the opacity changed event.

PsiOmegaDelta commented on code/modules/lighting/lighting_turf.dm:
```diff
@@ -1,37 +1,119 @@
-	if(obj && obj.opacity)
-		reconsider_lights()
+	var/old_counter = opaque_counter
+	if(Obj && Obj.opacity)
```
You want to register to the opacity event even if the object isn't blocking vision, seeing as that might at any time change.

PsiOmegaDelta commented on code/modules/lighting/lighting_turf.dm:
```diff
@@ -1,37 +1,119 @@
-		reconsider_lights()
+	var/old_counter = opaque_counter
+	if(Obj && Obj.opacity)
+		opaque_counter++
```
You probably want the opaque counter to be increased/decreased using 1 (or 2) proc(s) (also called by ```/turf/proc/handle_opacity_change```) and have it/them be responsible for calling reconsider light when going from 0 to 1 and 1 to 0.

As is you're spreading spreading around the responsibility of ```reconsider_lights()``` calls to multiple locations.

PsiOmegaDelta commented on code/modules/lighting/lighting_turf.dm:
```diff
@@ -1,37 +1,119 @@
+
+/turf/New()
+	if(opacity)
+		opaque_counter = opacity
```
You probably don't need the ```if(opacity)``` check, seeing how the opacity value should match the opaque_counter whether it's true or false.

PsiOmegaDelta replied:
You also want to do either ```opacity_set_event.register(src, src, /turf/proc/handle_opacity_change)``` or override ```/turf/set_opacity()``` to ensure the opacity flag updates when the turf itself changes.

In this particular case you probably want to do the latter, as I don't think BYOND would handle about half a million event registrations very well.

PsiOmegaDelta commented on code/datums/observation/opacity_set.dm:
```diff
@@ -17,11 +17,11 @@ var/decl/observ/opacity_set/opacity_set_event = new()
-	if(opacity != old_opacity)
-		opacity_set_event.raise_event(src, old_opacity, opacity)
+	if(new_opacity != old_opacity)
+		opacity = new_opacity
```
With the old parent set_opacity() no longer being a thing this should be ```opacity = !!new_opacity``` to ensure a TRUE/FALSE value (in spite of VV abuse or otherwise).

Datraen commented:
@chaoko99 https://github.com/Baystation12/Baystation12/pull/17044/files#diff-3301166f49df33c148321b06af346683R192

opacity_set_event was removed, as profiler showed that it used ~2x more cpu time than if checks and proc calls.

Datraen committed:
Tweaks lights

Corners seemed rather bright, adjusting by half the light power seemed to fix this.

StarSmasher44 commented:
https://github.com/tgstation/tgstation/pull/26975

/tg/ is always a great source for performance-related PR's, Bay could take advantage of that and catch up to the rest.

Real reason I posted that link is because if it isn't in yet and it seems compatible, I'd add it. Performance goes a long way for everyone.

Datraen commented:
I'll investigate compatability when I am free.

comma commented:
Porting from tg is pain since we don't have subsystems yet and a lot of things there are tied into those

StarSmasher44 commented:
Then subsystems could probably be merged soon-ish when someone is willing to review (and merge) it. And then this could be updated with the performance stuff everyone but us have (Apart from those running bay of course), and then this could be merged.

I guarantee it this will be great for Baystation and it's code.

Datraen commented:
![](http://i.imgur.com/e2ryo21.png)
Ready for review.

LonelyHowls commented:
Aaaaaaa we need this. It'll make the server look so much nicer, and make dark shit actually spooky.

comma reviewed (approved):
SS got merged, feel free to tackle a port in this PR or in a seprate.

Datraen commented:
Probably better to have the SS implementation tackled in a separate PR, otherwise this will be on hold for quite some time longer.
